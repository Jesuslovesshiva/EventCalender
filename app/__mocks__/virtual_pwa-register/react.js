import {useState} from 'react'

const useRegisterSW = jest.fn(() => {
  const [needRefresh, setNeedRefresh] = useState(true)

  return {
    needRefresh: [needRefresh, setNeedRefresh],
    updateServiceWorker: () => {}
  }
})

// eslint-disable-next-line import/prefer-default-export
export {useRegisterSW}
