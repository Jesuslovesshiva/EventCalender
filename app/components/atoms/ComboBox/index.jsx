import classNames from 'classnames'
import {omit} from 'lodash'
import noop from 'lodash/noop'
import PropTypes from 'prop-types'
import React, {useCallback, useState} from 'react'
import ReactSelect from 'react-select'

import useSyncState from '@/hooks/use-sync-state'

import ClearIndicator from './components/ClearIndicator'
import Control from './components/Control'
import DropdownIndicator from './components/DropdownIndicator'
import IndicatorsContainer from './components/IndicatorsContainer'
import Input from './components/Input'
import Menu from './components/Menu'
import MenuList from './components/MenuList'
import MultiValueContainer from './components/MultiValueContainer'
import MultiValueLabel from './components/MultiValueLabel'
import MultiValueRemove from './components/MultiValueRemove'
import Option from './components/Option'
import Placeholder from './components/Placeholder'
import SingleValue from './components/SingleValue'
import ValueContainer from './components/ValueContainer'

const useMenuIsOpen = props => {
  const [menuIsOpen, setMenuIsOpen] = useState(props.defaultMenuIsOpen)

  const events = {
    onKeyDown: useCallback(
      event => {
        if (event.key === 'Escape' && menuIsOpen) {
          event.stopPropagation()
          events.onMenuClose()
        }
        props.onKeyDown?.(event, menuIsOpen)
      },
      [menuIsOpen, setMenuIsOpen, props.onKeyDown]
    ),

    onMenuClose: useCallback(() => {
      setMenuIsOpen(false)
      props.onMenuClose?.()
    }, [setMenuIsOpen, props.onMenuClose]),

    onMenuOpen: useCallback(() => {
      setMenuIsOpen(true)
      props.onMenuOpen?.()
    }, [setMenuIsOpen, props.onMenuOpen])
  }

  return [menuIsOpen, events]
}

const ComboBox = React.forwardRef((props, ref) => {
  const [value, setValue] = useSyncState(props.value)
  const [options] = useSyncState(props.options)
  const [menuIsOpen, menuIsOpenEvents] = useMenuIsOpen(props)

  const onChange = next => {
    setValue(next)
    props.onChange(next)
  }

  const passThroughProps = omit(props, [
    'className',
    'components',
    'onChange',
    'onKeyDown',
    'onMenuClose',
    'onMenuOpen',
    'options',
    'value'
  ])

  return (
    <ReactSelect
      ref={ref}
      className={classNames('text-base w-full', props.className)}
      options={options}
      onChange={onChange}
      value={value}
      menuIsOpen={menuIsOpen}
      onKeyDown={menuIsOpenEvents.onKeyDown}
      onMenuOpen={menuIsOpenEvents.onMenuOpen}
      onMenuClose={menuIsOpenEvents.onMenuClose}
      components={{
        ClearIndicator,
        Control,
        DropdownIndicator,
        IndicatorSeparator: null,
        IndicatorsContainer,
        Input,
        Menu,
        MenuList,
        MultiValueContainer,
        MultiValueLabel,
        MultiValueRemove,
        Option,
        Placeholder,
        SingleValue,
        ValueContainer,
        ...props.components
      }}
      {...passThroughProps}
    />
  )
})

ComboBox.displayName = 'ComboBox'

const IDENTIFIER_PROP_TYPE = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.number
])
const OPTION = PropTypes.shape({
  label: PropTypes.string,
  node: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  value: IDENTIFIER_PROP_TYPE.isRequired
})
const COMBO_BOX_PROP_TYPES = {
  GROUPED_OPTION: PropTypes.shape({
    label: PropTypes.string.isRequired,
    options: PropTypes.arrayOf(OPTION)
  }),
  IDENTIFIER: IDENTIFIER_PROP_TYPE,
  OPTION
}

ComboBox.propTypes = {
  'aria-label': PropTypes.string,
  autoFocus: PropTypes.bool,
  className: PropTypes.string,
  components: PropTypes.object,
  getOptionLabel: PropTypes.func,
  hideSelectedOptions: PropTypes.bool,
  isDisabled: PropTypes.bool,
  isMulti: PropTypes.bool,
  onChange: PropTypes.func,
  options: PropTypes.oneOfType([
    PropTypes.arrayOf(COMBO_BOX_PROP_TYPES.OPTION),
    PropTypes.arrayOf(COMBO_BOX_PROP_TYPES.GROUPED_OPTION)
  ]),
  placeholder: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  value: PropTypes.oneOfType([
    COMBO_BOX_PROP_TYPES.OPTION,
    PropTypes.arrayOf(COMBO_BOX_PROP_TYPES.OPTION)
  ])
}

ComboBox.defaultProps = {
  'aria-label': null,
  autoFocus: false,
  className: null,
  components: {},
  // eslint-disable-next-line no-undefined
  getOptionLabel: undefined,
  hideSelectedOptions: false,
  isDisabled: false,
  isMulti: false,
  onChange: noop,
  options: [],
  placeholder: null,
  value: null
}

export {COMBO_BOX_PROP_TYPES}
export default ComboBox
