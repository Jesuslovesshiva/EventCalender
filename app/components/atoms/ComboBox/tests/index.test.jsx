import {fireEvent} from '@testing-library/react'

import ComboBox from '@/components/atoms/ComboBox'
import render from '@/utils/test-utils/custom-render'

const defaultProps = {
  options: [
    {label: 'Option #1', value: 'option-1'},
    {isDisabled: true, label: 'Option #2', value: 'option-2'},
    {label: 'Option #3', value: 'option-3'}
  ]
}

const renderComponent = (props = {}) =>
  render(<ComboBox {...defaultProps} {...props} />)

describe('ComboBox', () => {
  it('should match the snapshot', () => {
    const {container} = renderComponent()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should match the snapshot if the menu is open', () => {
    const {container, getByRole} = renderComponent()
    const combobox = getByRole('combobox')
    fireEvent.input(combobox, {target: {value: 'O'}})
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should match the snapshot if an option has been selected', () => {
    const {container, queryByText, getByRole} = renderComponent()
    const combobox = getByRole('combobox')
    fireEvent.input(combobox, {target: {value: 'O'}})
    fireEvent.click(queryByText(defaultProps.options[0].label))
    expect(container.firstChild).toMatchSnapshot()
  })

  describe('isMulti', () => {
    it('should match the snapshot', () => {
      const {container, getByRole} = renderComponent({isMulti: true})
      const combobox = getByRole('combobox')
      fireEvent.input(combobox, {target: {value: 'O'}})
      expect(container.firstChild).toMatchSnapshot()
    })

    it('should match the snapshot if an option has been selected', () => {
      const {container, queryByText, getByRole} = renderComponent({
        isMulti: true
      })
      const combobox = getByRole('combobox')
      fireEvent.input(combobox, {target: {value: 'O'}})
      fireEvent.click(queryByText(defaultProps.options[0].label))
      expect(container.firstChild).toMatchSnapshot()
    })

    it('should call the on change prop and close the menu if a value has been selected', () => {
      const onChange = jest.fn()
      const {queryByText, getByRole} = renderComponent({onChange})
      const combobox = getByRole('combobox')
      fireEvent.input(combobox, {target: {value: 'O'}})
      fireEvent.click(queryByText(defaultProps.options[0].label))
      expect(onChange).toHaveBeenCalledTimes(1)
      expect(onChange).toHaveBeenCalledWith(defaultProps.options[0])
      fireEvent.input(combobox, {target: {value: 'O'}})
      fireEvent.click(queryByText(defaultProps.options[2].label))
      expect(onChange).toHaveBeenCalledWith(defaultProps.options[2])
    })
  })

  it('should render the passed options if open', () => {
    const {getByText, getByRole} = renderComponent()
    const combobox = getByRole('combobox')
    fireEvent.input(combobox, {target: {value: 'O'}})
    defaultProps.options.forEach(option =>
      expect(getByText(option.label)).toBeInTheDocument()
    )
  })

  it('should call the on change prop and close the menu if a value has been selected', () => {
    const onChange = jest.fn()
    const {queryByText, getByRole} = renderComponent({onChange})
    const combobox = getByRole('combobox')
    fireEvent.input(combobox, {target: {value: 'O'}})
    fireEvent.click(queryByText(defaultProps.options[0].label))
    expect(queryByText(defaultProps.options[0].label)).toBeInTheDocument()
    expect(queryByText(defaultProps.options[1].label)).not.toBeInTheDocument()
    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenCalledWith(defaultProps.options[0])
  })

  it('should not close the menu if the option to be selected is disabled', () => {
    const {queryByText, getByRole} = renderComponent({})
    const combobox = getByRole('combobox')
    fireEvent.input(combobox, {target: {value: 'O'}})
    fireEvent.click(queryByText(defaultProps.options[1].label))
    defaultProps.options.forEach(option =>
      expect(queryByText(option.label)).toBeInTheDocument()
    )
  })

  it('should close the menu on escape', () => {
    const {queryByText, getByRole} = renderComponent({})
    const combobox = getByRole('combobox')
    fireEvent.keyDown(combobox, {key: 'ArrowDown'})
    expect(queryByText(defaultProps.options[1].label)).toBeInTheDocument()
    fireEvent.keyDown(combobox, {key: 'Escape'})
    expect(queryByText(defaultProps.options[1].label)).not.toBeInTheDocument()
  })
})
