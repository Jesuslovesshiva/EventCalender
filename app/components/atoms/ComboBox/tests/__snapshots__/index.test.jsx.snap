// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboBox isMulti should match the snapshot 1`] = `
<div
  class="text-base w-full css-b62m3t-container"
>
  <span
    class="css-1f43avz-a11yText-A11yText"
    id="react-select-5-live-region"
  />
  <span
    aria-atomic="false"
    aria-live="polite"
    aria-relevant="additions text"
    class="css-1f43avz-a11yText-A11yText"
  />
  <div
    class="!min-h-0 !rounded-1 !transition-shadow pl-3 gap-1.5 !bg-transparent !border-gray-medium hover:!border-surface hover:!bg-silver hover:!shadow-md !rounded-b-none h-[28px] pl-2.5 css-1s2u09g-control"
  >
    <div
      class="!py-0 !px-0 h-fit css-319lph-ValueContainer"
    >
      <div
        class="!m-0 !p-0 !text-gray-dark css-1pndypt-Input"
        data-value="O"
      >
        <input
          aria-autocomplete="list"
          aria-controls="react-select-5-listbox"
          aria-describedby="react-select-5-placeholder"
          aria-expanded="true"
          aria-haspopup="true"
          aria-owns="react-select-5-listbox"
          autocapitalize="none"
          autocomplete="off"
          autocorrect="off"
          class=""
          id="react-select-5-input"
          role="combobox"
          spellcheck="false"
          style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
          tabindex="0"
          type="text"
          value="O"
        />
      </div>
    </div>
    <div
      class="h-5 m-auto css-1hb7zxy-IndicatorsContainer"
    >
      <div
        aria-hidden="true"
        class="text-2xl !px-3 !py-1 !text-gray-medium rotate-180 css-tlfecz-indicatorContainer"
      >
        <div
          class="fill-current w-[1em] h-[1em] inline-block"
          data-icon="arrow_drop_down"
        />
      </div>
    </div>
  </div>
  <div
    class="!m-0 !rounded-t-none !shadow css-26l3qy-menu"
    id="react-select-5-listbox"
  >
    <div
      class="!p-0 css-4ljt47-MenuList"
    >
      <div
        aria-disabled="false"
        class="!py-1.5 !px-3 border-b border-b-divider/30 last:border-b-0 last:rounded-b !bg-silver !text-gray-dark css-1n7v3ny-option"
        id="react-select-5-option-0"
        tabindex="-1"
      >
        Option #1
      </div>
      <div
        aria-disabled="true"
        class="!py-1.5 !px-3 border-b border-b-divider/30 last:border-b-0 last:rounded-b !text-gray-medium css-165luzk-option"
        id="react-select-5-option-1"
        tabindex="-1"
      >
        Option #2
      </div>
      <div
        aria-disabled="false"
        class="!py-1.5 !px-3 border-b border-b-divider/30 last:border-b-0 last:rounded-b css-yt9ioa-option"
        id="react-select-5-option-2"
        tabindex="-1"
      >
        Option #3
      </div>
    </div>
  </div>
</div>
`;

exports[`ComboBox isMulti should match the snapshot if an option has been selected 1`] = `
<div
  class="text-base w-full css-b62m3t-container"
>
  <span
    class="css-1f43avz-a11yText-A11yText"
    id="react-select-6-live-region"
  />
  <span
    aria-atomic="false"
    aria-live="polite"
    aria-relevant="additions text"
    class="css-1f43avz-a11yText-A11yText"
  />
  <div
    class="!min-h-0 !rounded-1 !transition-shadow pl-3 gap-1.5 !bg-transparent !border-gray-medium hover:!border-surface hover:!bg-silver hover:!shadow-md h-[28px] pl-2.5 css-1s2u09g-control"
  >
    <div
      class="!py-0 !px-0 h-fit css-g1d714-ValueContainer"
    >
      <div
        class="css-1rhbuit-multiValue h-[18px] !bg-transparent border border-gray-medium !rounded !ml-0 !mr-[3px] !my-0"
      >
        <div
          class="css-12jo7m5 px-[5px] !py-0 inline-flex items-center"
        >
          Option #1
        </div>
        <div
          aria-label="Remove Option #1"
          class="css-xb97g8 hover:!bg-transparent !pl-0 text-gray-medium hover:!text-gray-dark/50"
          role="button"
        >
          <div
            class="fill-current w-[1em] h-[1em] inline-block"
            data-icon="close"
          />
        </div>
      </div>
      <div
        class="!m-0 !p-0 !text-gray-dark css-6j8wv5-Input"
        data-value=""
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-haspopup="true"
          autocapitalize="none"
          autocomplete="off"
          autocorrect="off"
          class=""
          id="react-select-6-input"
          role="combobox"
          spellcheck="false"
          style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
          tabindex="0"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="h-5 m-auto css-1hb7zxy-IndicatorsContainer"
    >
      <div
        aria-hidden="true"
        class="text-2xl !py-1 text-gray-medium hover:!text-gray-dark/50 hover:cursor-pointer !px-0 css-tlfecz-indicatorContainer"
      >
        <div
          class="fill-current w-[1em] h-[1em] inline-block"
          data-icon="close"
        />
      </div>
      <div
        aria-hidden="true"
        class="text-2xl !px-3 !py-1 !text-gray-medium css-tlfecz-indicatorContainer"
      >
        <div
          class="fill-current w-[1em] h-[1em] inline-block"
          data-icon="arrow_drop_down"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ComboBox should match the snapshot 1`] = `
<div
  class="text-base w-full css-b62m3t-container"
>
  <span
    class="css-1f43avz-a11yText-A11yText"
    id="react-select-2-live-region"
  />
  <span
    aria-atomic="false"
    aria-live="polite"
    aria-relevant="additions text"
    class="css-1f43avz-a11yText-A11yText"
  />
  <div
    class="!min-h-0 !rounded-1 !transition-shadow pl-3 gap-1.5 !bg-transparent !border-gray-medium hover:!border-surface hover:!bg-silver hover:!shadow-md css-1s2u09g-control"
  >
    <div
      class="!py-0 !px-0 css-319lph-ValueContainer"
    >
      <div
        class="!m-0 specific:text-gray-medium line-clamp-1 css-14el2xx-placeholder"
        id="react-select-2-placeholder"
      />
      <div
        class="!m-0 !p-0 !text-gray-dark css-6j8wv5-Input"
        data-value=""
      >
        <input
          aria-autocomplete="list"
          aria-describedby="react-select-2-placeholder"
          aria-expanded="false"
          aria-haspopup="true"
          autocapitalize="none"
          autocomplete="off"
          autocorrect="off"
          class=""
          id="react-select-2-input"
          role="combobox"
          spellcheck="false"
          style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
          tabindex="0"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class=" css-1hb7zxy-IndicatorsContainer"
    >
      <div
        aria-hidden="true"
        class="text-2xl !px-3 !py-1 !text-gray-medium css-tlfecz-indicatorContainer"
      >
        <div
          class="fill-current w-[1em] h-[1em] inline-block"
          data-icon="arrow_drop_down"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ComboBox should match the snapshot if an option has been selected 1`] = `
<div
  class="text-base w-full css-b62m3t-container"
>
  <span
    class="css-1f43avz-a11yText-A11yText"
    id="react-select-4-live-region"
  />
  <span
    aria-atomic="false"
    aria-live="polite"
    aria-relevant="additions text"
    class="css-1f43avz-a11yText-A11yText"
  />
  <div
    class="!min-h-0 !rounded-1 !transition-shadow pl-3 gap-1.5 !bg-transparent !border-gray-medium hover:!border-surface hover:!bg-silver hover:!shadow-md css-1s2u09g-control"
  >
    <div
      class="!py-0 !px-0 css-319lph-ValueContainer"
    >
      <div
        class="!m-0 !p-0 !text-gray-dark css-qc6sy-singleValue"
      >
        Option #1
      </div>
      <div
        class="!m-0 !p-0 !text-gray-dark css-6j8wv5-Input"
        data-value=""
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-haspopup="true"
          autocapitalize="none"
          autocomplete="off"
          autocorrect="off"
          class=""
          id="react-select-4-input"
          role="combobox"
          spellcheck="false"
          style="opacity: 0; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
          tabindex="0"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class=" css-1hb7zxy-IndicatorsContainer"
    >
      <div
        aria-hidden="true"
        class="text-2xl !px-3 !py-1 !text-gray-medium css-tlfecz-indicatorContainer"
      >
        <div
          class="fill-current w-[1em] h-[1em] inline-block"
          data-icon="arrow_drop_down"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ComboBox should match the snapshot if the menu is open 1`] = `
<div
  class="text-base w-full css-b62m3t-container"
>
  <span
    class="css-1f43avz-a11yText-A11yText"
    id="react-select-3-live-region"
  />
  <span
    aria-atomic="false"
    aria-live="polite"
    aria-relevant="additions text"
    class="css-1f43avz-a11yText-A11yText"
  />
  <div
    class="!min-h-0 !rounded-1 !transition-shadow pl-3 gap-1.5 !bg-transparent !border-gray-medium hover:!border-surface hover:!bg-silver hover:!shadow-md !rounded-b-none css-1s2u09g-control"
  >
    <div
      class="!py-0 !px-0 css-319lph-ValueContainer"
    >
      <div
        class="!m-0 !p-0 !text-gray-dark css-1pndypt-Input"
        data-value="O"
      >
        <input
          aria-autocomplete="list"
          aria-controls="react-select-3-listbox"
          aria-describedby="react-select-3-placeholder"
          aria-expanded="true"
          aria-haspopup="true"
          aria-owns="react-select-3-listbox"
          autocapitalize="none"
          autocomplete="off"
          autocorrect="off"
          class=""
          id="react-select-3-input"
          role="combobox"
          spellcheck="false"
          style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
          tabindex="0"
          type="text"
          value="O"
        />
      </div>
    </div>
    <div
      class=" css-1hb7zxy-IndicatorsContainer"
    >
      <div
        aria-hidden="true"
        class="text-2xl !px-3 !py-1 !text-gray-medium rotate-180 css-tlfecz-indicatorContainer"
      >
        <div
          class="fill-current w-[1em] h-[1em] inline-block"
          data-icon="arrow_drop_down"
        />
      </div>
    </div>
  </div>
  <div
    class="!m-0 !rounded-t-none !shadow css-26l3qy-menu"
    id="react-select-3-listbox"
  >
    <div
      class="!p-0 css-4ljt47-MenuList"
    >
      <div
        aria-disabled="false"
        class="!py-1.5 !px-3 border-b border-b-divider/30 last:border-b-0 last:rounded-b !bg-silver !text-gray-dark css-1n7v3ny-option"
        id="react-select-3-option-0"
        tabindex="-1"
      >
        Option #1
      </div>
      <div
        aria-disabled="true"
        class="!py-1.5 !px-3 border-b border-b-divider/30 last:border-b-0 last:rounded-b !text-gray-medium css-165luzk-option"
        id="react-select-3-option-1"
        tabindex="-1"
      >
        Option #2
      </div>
      <div
        aria-disabled="false"
        class="!py-1.5 !px-3 border-b border-b-divider/30 last:border-b-0 last:rounded-b css-yt9ioa-option"
        id="react-select-3-option-2"
        tabindex="-1"
      >
        Option #3
      </div>
    </div>
  </div>
</div>
`;
